### PR TITLE
DYN-5993: Update avalonedit, migrate dynamoservicestests, pythonnodemodelswpf

### PR DESCRIPTION
**Pull request by aparajit-pratap**
### Purpose

Update avalonedit
migrate:
 dynamoservicestests, pythonnodemodelswpf

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB 

### Release Notes

(FILL ME IN) Brief description of the fix / enhancement. **Mandatory section** 

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ namespace PythonNodeModelsWpf
     /// Allows producing tab based foldings
     /// https://stackoverflow.com/questions/47224064/avalonedit-foldingstrategy-by-indent-python/47577500#47577500
     /// </summary>
-    internal class TabFoldingStrategy : AbstractFoldingStrategy
+    internal class TabFoldingStrategy : XmlFoldingStrategy
```
Comment (aparajit-pratap) [inline src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs:13]: `AbstractFoldingStrategy `has been removed from AvalonEdit versions > 4.x

File: LICENSE.txt [hunk 2]
```diff
@@ -127,26 +127,19 @@ The above copyright notice and this permission notice shall be included in all c
...
 
-Portions related to the Avalon Edit v.4.3.1.9429 are © 2014 AlphaSierraPapa for the SharpDevelop Team. All rights reserved. Avalon Edit is licensed under the GNU Lesser General Public License v.3.0, which can be found at http://opensource.org/licenses/lgpl-3.0.html. A text copy of this license is included on the media provided by Autodesk or with the download of this Autodesk software. You may obtain a copy of the source code for Avalon Edit v.4.3.1.9429 from www.autodesk.com/lgplsource or by sending a written request to:
+MIT License
 
-Autodesk, Inc.
-Attention: General Counsel
-Legal Department
-111 McInnis Parkway
-San Rafael, CA 9490
+Copyright (c) <year> <copyright holders>
```
Comment (mjkkirschner) [inline LICENSE.txt:136]: ??

**Review by mjkkirschner — CHANGES_REQUESTED** [review]
looks like some actual test failures:
https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/9608/

and just a few questions

File: test/Libraries/IronPythonTests/IronPythonTests.csproj [hunk 3]
```diff
@@ -12,9 +12,9 @@
     <AssemblyName>IronPythonTests</AssemblyName>
   </PropertyGroup>
   <ItemGroup>
-	  <PackageReference Include="AvalonEdit" Version="4.3.1.9430" />
+	  <PackageReference Include="AvalonEdit" Version="6.3.0.90"/>
```
[the remark this answers is not in the document]
Comment (aparajit-pratap) [reply test/Libraries/IronPythonTests/IronPythonTests.csproj:15]: Good question, I'll check that.

[the remark this answers is not in the document]
Comment (aparajit-pratap) [reply test/Libraries/IronPythonTests/IronPythonTests.csproj:15]: Actually, I don't see Revit using AvalonEdit directly. I think you mean Revit AddIns?

File: src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs [hunk 4]
```diff
@@ -25,7 +25,7 @@ internal TabFoldingStrategy()
         /// <summary>
         /// Create <see cref="NewFolding"/>s for the specified document.
         /// </summary>
-        public override IEnumerable<NewFolding> CreateNewFoldings(TextDocument document, out int firstErrorOffset)
+        public new IEnumerable<NewFolding> CreateNewFoldings(TextDocument document, out int firstErrorOffset)
```
[the remark this answers is not in the document]
Comment (mjkkirschner) [reply src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs:28]: okay, got it, so `AbstractFoldingStrategy` is gone and there is no similar base class replacement given so you had to hide the old method.

Do we actually need this new method, though? Does the base class implementation not just work?

File: test/Libraries/IronPythonTests/IronPythonTests.csproj (hunk 3 above)
[the remark this answers is not in the document]
Comment (aparajit-pratap) [reply test/Libraries/IronPythonTests/IronPythonTests.csproj:15]: Leaning toward going ahead with this version for now. I checked with Revit and they're ok with it - see slack discussion [here](https://autodesk.slack.com/archives/C0531NMR189/p1689005094202889).

File: src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs [hunk 5]
```diff
@@ -10,7 +10,7 @@ namespace PythonNodeModelsWpf
     /// Allows producing tab based foldings
     /// https://stackoverflow.com/questions/47224064/avalonedit-foldingstrategy-by-indent-python/47577500#47577500
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs:11]: is this comment still valid?

File: test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs [hunk 6]
```diff
@@ -145,16 +147,15 @@ public void OnTextAreaTextEnteringEventTest()
...
-            textArea.RaiseEvent(new KeyEventArgs(
-                Keyboard.PrimaryDevice,
-                PresentationSource.FromVisual(codeEditor),
-                0,
-                Key.End)
-            {
-                RoutedEvent = Keyboard.KeyDownEvent
-            }
-            );
+            var textArea = scriptWindow.editText.TextArea;
```
Comment (mjkkirschner) [inline test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs:150]: is there any possibility that this any of these are null? Can we use `?` operator?

File: test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs [hunk 7]
```diff
@@ -145,16 +147,15 @@ public void OnTextAreaTextEnteringEventTest()
...
+
+            // Get the current caret position
+            TextLocation caretPosition = textArea.Caret.Location;
+
+            // Set the caret position to the end of the line
+            var endOfLine = new TextLocation(caretPosition.Line,
+                codeEditor.Document.GetLineByNumber(caretPosition.Line).Length + 1);
```
Comment (mjkkirschner) [inline test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs:157]: possible this could return null?

**Review by mjkkirschner — APPROVED** [review]
LGTM - some APIs I am not familiar with though so maybe best to check for null / use null coalescing operator?

**Comment by mjkkirschner** [comment]
@aparajit-pratap looks like some conflicts.

File: src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs (hunk 5 above)
Comment (aparajit-pratap) [reply src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs:11]: Yes, this code is still largely lifted from there.

File: test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs (hunk 6 above)
Comment (aparajit-pratap) [reply test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs:150]: This is a test. Do you still feel that some defensive coding is necessary here?

Comment (aparajit-pratap) [reply test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs:150]: For some reason, the previous logic of raising the event started failing with the new package. I couldn't figure out why so I used this other way of moving the cursor position to the end of the line for the purpose of this test.

**Pull request merged**